### ISSUE TITLE
problem with memgraph using neo4j driver.net

### ISSUE THREAD

**Issue by imandoostie** [memgraph/memgraph#1373]
i'm using foreach to read from csv file and create edge between tow nodes in Memgraph which is too slow to say the least
like 2000 edge in 45 minutes
then i switch my foreach to parralel.foreach
that's when i start to get this error 
`Cannot resolve conflicting transactions. You can retry this transaction when the conflicting transaction is finished'
then i tried to switch my query to use BEGIN and COMMIT which i tried to in Memgraph lab and it worked but not in neo4j .NET driver
any suggestions? 
also i know i could try to read csv from Memgraph lab but i have to do it one by one which take forever to finish

**Comment by antejavor** [comment]
Hi @imandoostie, these are very low numbers. Have you tried setting up the property indexes on the nodes you are matching? 
```
LOAD CSV FROM "/path-to/people_relationships.csv" WITH HEADER AS row
MATCH (p1:Person {id: row.id_from}), (p2:Person {id: row.id_to})
CREATE (p1)-[:IS_FRIENDS_WITH]->(p2);
```

**Comment by imandoostie** [comment]
hi @antejavor 
thanks for your reply
i tried this command , what i have in my csv file is about 10 million relation and if i use this command on my server
which is (ubuntu , 8core , 64gb ram, 250gb SSD)
by my measurement it's going to take 150 days to finish 
i'm looking for an alternative way if it's possible .. please

**Comment by antejavor** [comment]
Sorry, @imandoostie I forgot to add this line to the first comment: `CREATE INDEX ON :Person(id);`

This way, an index will be used. The total load time will be a few minutes for your scale. You can also switch to analytical mode. 

I guess you didn't create and index? 

Read this in detail: https://memgraph.com/docs/data-migration/csv#increase-import-speed

**Comment by imandoostie** [comment]
here is my code 

```
var x = new CancellationToken();
await Parallel.ForEachAsync<DataRow>(csvData.AsEnumerable(), async (item, x) =>
{

    string query = "" +
                    "MATCH (p1{_id:" + "'" + item[0] + "'" + "}),(p2{_id: " + "'" + item[1] + "'" + "}) " +
                   "CREATE (p1)-[r:" + item[2] + "]->(p2) " +
                   "RETURN r" ;
    await Task.Run(() => InsertDB(query));
});


async static Task InsertDB(string query)
{
    using var _driver = GraphDatabase.Driver("bolt://127.0.0.1:7687", AuthTokens.None);
    using var session = _driver.AsyncSession(x => x.WithDefaultAccessMode(AccessMode.Write));
    await session.ExecuteWriteAsync(async tx =>
    {
        var reader = await tx. RunAsync(query);
    });
}
```

`var reader = await tx. RunAsync(query);`

this is where i get my error every time

**Comment by imandoostie** [comment]
@katarinasupe 
hello and thanks for your reply 
i did what you asked me and added index on each one of my node and change my query to using label in both nodes and relations
the result was not promising , what i had before was going take 150 days to finish 
now it may finish in 75 days
i can't use in memory storage mode not an option
how much do i have to wait for this conflict transaction issue to be fixed?
also i'm open to any other suggestions

**Comment by katarinasupe** [comment]
Thanks for trying the indexes @imandoostie, I suspect there's some small detail that's making this so slow but it can be hard to debug remotely without seeing exactly what's happening. This definitely isn't supposed to take this long, though.

If you want, we can [hop on a call](https://memgraph.com/office-hours) and debug this together, I think we'd be able to fix this pretty quickly. Or, if you are able to wait for some I can try to setup a .net test on my machine to try to replicate this.

**Comment by as51340** [comment]
hi @imandoostie, would it be okay that you share a sample dataset you are using with us in order to more easily reproduce the issue? These are too low numbers for importing speed, let's try to fix that!

**Comment by as51340** [comment]
And are you maybe using disk storage? If that is the case, we are currently working on one of our PRs to speed up load csv.

**Comment by imandoostie** [comment]
@as51340 hello 
yes i'm using ssd disk storage

**Comment by as51340** [comment]
Just to double check: I meant to ask whether you are using disk storage mode in Memgraph?

**Comment by imandoostie** [comment]
@as51340 yes i'm using disk storage mode, i think one of the reasons i'm getting conflict transaction error is because i'm using disk storage mode, but unfortunately  ram mode isn't an option here

**Comment by as51340** [comment]
okay in that case, [this PR](https://github.com/memgraph/memgraph/pull/1332) should solve the importing speed of load csv command. I suggest you don't use parallel foreach since it is not polished for disk storage and for importing 2000 edges, the single-threaded load csv command will be more than enough.

May I ask why RAM mode isn't an option?

**Comment by katarinasupe** [comment]
Hi @imandoostie, any update on the issue you're experiencing?
